idx/memory/memory.go

```diff
@@ -0,0 +1,475 @@
+package memory
+
+import (
+	"flag"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/raintank/met"
+	"github.com/raintank/metrictank/idx"
+	"github.com/raintank/worldping-api/pkg/log"
+	"github.com/rakyll/globalconf"
+	"gopkg.in/raintank/schema.v1"
+)
+
+var (
+	idxOk             met.Count
+	idxFail           met.Count
+	idxAddDuration    met.Timer
+	idxGetDuration    met.Timer
+	idxListDuration   met.Timer
+	idxFindDuration   met.Timer
+	idxDeleteDuration met.Timer
+
+	Enabled bool
+)
+
+func ConfigSetup() {
+	memoryIdx := flag.NewFlagSet("memory-idx", flag.ExitOnError)
+	memoryIdx.BoolVar(&Enabled, "enabled", false, "")
+	globalconf.Register("memory-idx", memoryIdx)
+}
+
+type Tree struct {
+	Items map[string]*Node
+}
+
+type Node struct {
+	Path     string
+	Children []string
+	Leaf     bool
+}
+
+func (n *Node) String() string {
+	t := "branch"
+	if n.Leaf {
+		t = "leaf"
+	}
+	return fmt.Sprintf("%s - %s", t, n.Path)
+}
+
+// Implements the the "MetricIndex" interface
+type MemoryIdx struct {
+	sync.Mutex
+	DefById map[string]*schema.MetricDefinition
+	Tree    map[int]*Tree
```

Comment:
That's a good point about map vs slice.

I extended your program in two ways:
1) do similar test for list. it shows that for lists, nopointers is a bit faster (and lists in general are faster than maps)
2) handle b.N > 5M more fairly and make sure the go bench runner can always converge. (the math is not exact but it doesn't have to be)

otherwise it pretty much confirms your findings. question of course is how these measurements stack up against the other concerns (GC overhead, memory usage, etc)

so why did you decide to go pointerless in fb8daaf ?